python-lib/dku_email_client.py

```diff
@@ -0,0 +1,112 @@
+
+        logging.info(f"Configured channel messaging client with channel {channel_id} - type: {self.channel.type}, "
+                     f"sender: {self.channel.sender}, plain_text? {self.plain_text}")
+
+    def send_email_impl(self, sender, recipient, email_subject, email_body, attachment_files):
+        """
+        :param sender: sender email, str - is ignored if a sender configured for the channel
+        :param recipient: recipient email, str
+        :param email_subject: str
+        :param email_body: body of either plain text or html, str
+        :param attachment_files:attachments as list of  AttachmentFile
+        """
+        files = [(a.file_name, a.data, f"{a.mime_type}/{a.mime_subtype}") for a in attachment_files]
+
+        sender_to_use = None if self.channel.sender else sender
+        self.channel.send(self.project_id, [recipient], email_subject, email_body, attachments=files, plain_text=self.plain_text, sender=sender_to_use)
+
+    def quit(self):
+        pass
+
+
+class SmtpEmailClient(AbstractMessageClient):
+    """ Client for sending email - direct SMTP implementation """
+
+    def __init__(self, plain_text, smtp_config):
+        super().__init__(plain_text)
+        self.smtp = smtplib.SMTP(smtp_config.smtp_host, port=smtp_config.smtp_port)
+        # Use TLS if set
+        if smtp_config.smtp_use_tls:
+            self.smtp.starttls()
+        # Use credentials if set
+        if smtp_config.smtp_use_auth:
+            self.smtp.login(str(smtp_config.smtp_user), str(smtp_config.smtp_pass))
+        logging.info(f"Configured an STMP mail client with host: {smtp_config.smtp_host}, port: {smtp_config.smtp_port}, "
+                     f"tls? {smtp_config.smtp_use_tls}, auth? {smtp_config.smtp_use_auth}, plain_text? {self.plain_text}")
+
+    def send_email_impl(self, sender, recipient, email_subject, email_body, attachment_files):
+        msg = MIMEMultipart()
+        msg["From"] = sender
+        msg["To"] = recipient
+        msg["Subject"] = email_subject
+        body_encoding = "utf-8"
+        text_type = 'plain' if self.plain_text else 'html'
+        msg.attach(MIMEText(email_body, text_type, body_encoding))
+        for attachment_file in attachment_files:
+            if attachment_file.mime_type == "application":
+                mime_app = MIMEApplication(attachment_file.data, _subtype=attachment_file.mime_subtype)
+            elif attachment_file.mime_type == "text":
+                mime_app = MIMEText(attachment_file.data, _subtype=attachment_file.mime_subtype, _charset="utf-8")
+            else:
+                raise Exception(f'Cannot handle mime type {attachment_file.mime_type}')
+            mime_app.add_header("Content-Disposition", 'attachment', filename=attachment_file.file_name)
+            msg.attach(mime_app)
+        self.smtp.sendmail(sender, [recipient], msg.as_string())
+
+    def quit(self):
+        """ Do any disconnection needed"""
+        self.smtp.quit()
```

Comment:
Should be also declared in the abstract class

Reply:
Done https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/ee8f64a5f31f04ab7a85974561e6dcee19aa45ad